src/js/classes/BackgroundFetch.js

```diff
@@ -0,0 +1,84 @@
+import getURLsForDownload from '../utils/getURLsForDownload';
+
+const BG_FETCH_ID_TEMPLATE = /kino-(?<videoId>[a-z-_]+)/;
+
+export default class BackgroundFetch {
+  /**
+   * Construct a new Background Fetch wrapper.
+   */
+  constructor() {
+    this.id = '';
+    this.videoId = '';
+    this.onprogress = () => {};
+    this.ondone = () => {};
+  }
+
+  async maybeAbort(swReg) {
+    const existingBgFetch = await swReg.backgroundFetch.get(this.id);
+    if (existingBgFetch) await existingBgFetch.abort();
+  }
+
+  /**
+   *
+   * @param {BackgroundFetchRegistration} registration Background Fetch Registration object.
+   */
+  fromRegistration(registration) {
+    const matches = registration.id.match(BG_FETCH_ID_TEMPLATE);
+
+    this.id = registration.id;
+    this.videoId = matches.groups?.videoId || '';
+  }
+
+  async start(videoData) {
+    this.videoId = videoData.id;
+    this.id = `kino-${this.videoId}`;
+
+    const urls = await getURLsForDownload(this.videoId);
+
+    /** @type {Promise<Response[]>} */
+    const requests = urls.map((url) => fetch(url, { method: 'HEAD' }));
+
+    Promise.all(requests).then((responses) => {
+      const sizes = responses.map((response) => response.headers.get('Content-Length'));
+      const downloadTotal = sizes.includes(null)
+        ? null
+        : sizes.reduce((total, size) => total + parseInt(size, 10), 0);
+
+      // eslint-disable-next-line compat/compat
+      navigator.serviceWorker.ready.then(async (swReg) => {
+        this.maybeAbort(swReg);
+
+        /** @type {BackgroundFetchRegistration} */
+        const bgFetch = await swReg.backgroundFetch.fetch(
+          this.id,
+          urls,
+          {
+            title: videoData.title,
```

Comment:
`videoData.title` may end up being confusing. See video at http://localhost:5000/multiple-sources/ for instance.
Shall we have something like `Downloading "${videoData.title}" video`? ([Source](https://github.com/WICG/background-fetch#:~:text=title%20%2D%20Something%20descriptive%20to%20show%20in%20UI%2C%20such%20as%20%22Uploading%20%27Holiday%20in%20Rome%27%22%20or%20%22Downloading%20%27Catastrophe%20season%202%20episode%201%27%22.))

![image](https://user-images.githubusercontent.com/634478/142875248-41e7957d-2ad1-4e39-bad1-82053c5358e8.png)

Reply:
Updated in 7979e32.